_posts/2025-04-22-test-classloading-rewrite.adoc

```diff
@@ -0,0 +1,60 @@
+---
+layout: post
+title: 'The internals (and a few externals) of Quarkus test classloading have changed'
+tags: announcement testing
+synopsis: 'The way that Quarkus loads test classes has been updated. Most tests will not need to change, but here are some things to watch out for.'
+author: hcummins
+---
+
+== What's changing?
+
+The internals of Quarkus test classloading have been rewritten in 3.22.
+In particular, how `@QuarkusTest` test classes are loaded is changing.
```

Comment:
```suggestion
In particular, this is not affecting Quarkus production classloaders it is only changing how `@QuarkusTest` test classes are loaded during tests.
```

assuming that is a true statement its worth mentioning this is only affecting tests, not prod.

...how about devmode? is it affected ?

Reply:
Dev mode isn't changed either. It is only `@QuarkusTest` (not even other test types). I'll add a clarifying sentence to list the exclusions explicitly.

Reply:
ok, so that devservices start in a different way is *only* for tests not devmode?

Reply:
Technically, they start in the same way as they've always done (in the augmentation phase), they just all do it at the same time. :) 

But flippancy aside, I'm not aware of any interactions between these changes and dev mode. "Dev mode" is a bit overloaded because we use it to mean both `quarkus:dev`, which actually launches two applications, and also the "not-test" application part of that. So in `quarkus:dev` the tests will get run in a different way, but the application side should be the same. I wondered if the frontloaded dev services could conflict with the application-side-dev-service, but I don't think that's an issue, because if there was a possibility for conflict on port numbers, it would have already manifested halfway through a test run, in the old architecture.

Reply:
No, the change only affects test, not dev. However, it affects continuous testing happening in dev mode.

Reply:
I would write it using multiple sentences:

The internals of Quarkus test classloading have been rewritten in 3.22.
It does **not** affect production and dev modes. However, it changes how `@QuarkusTest`, `@QuarkusIntegrationTest`, and other Quarkus tests are loaded.

Reply:
I've gone with the last one, as suggested.